### ISSUE TITLE
Numerical error codes for exceptions in Deno (like HTTP status codes or custom numbers)

### ISSUE THREAD

**Issue by rsp** [denoland/deno#3386]
**Update:** To make it clear, I mean a convention like tsc has that lets us reference error messages like `TS2304` in issue #3391 and google it easily, not to parse the strings programatically but more for good user experience.

### Idea:

Using numerical error codes for error messages in Deno.

Maybe using standard HTTP status code numbers in error messages in Deno where it makes sense, or maybe having our own list of error codes.

This is not to replace `Error` classes to identify exceptions inside of a program, just for the messages printed to the user.

### Context:

Recently I was writing tests in Deno to test for exceptions with `assertThrows()` from [std/testing/asserts.ts](https://github.com/denoland/deno/blob/master/std/testing/asserts.ts) which had a `msgIncludes` parameter that I used to match the error. 

On every OS Deno throwed an error with message "No such file" except for Windows, where it was: "The system cannot find the file specified. (os error 2)".

Of course matching specific error messages is never a good idea, maybe it can break in the future when we have localized messages or they just change to be more descriptive etc. but since Deno is using good practices from the Web, then maybe it would be a good idea to use HTTP status codes for errors like this for better user experience.

For example:

- "404 No such file: x.txt"
- "404 Not found: x.txt"
- "404 x.txt - The system cannot find the file specified. (os error 2)"
- "404 x.txt - sorry but I cannot load configuration"

would all be obvious to anyone who doesn't even know English.

If HTTP is not enough or would be confusing, then maybe something like this:

- "ERR1234 No such file: x.txt"
- "ERR1234 Not found: x.txt"
- "ERR1234 x.txt - The system cannot find the file specified. (os error 2)"
- "ERR1234 x.txt - sorry but I cannot load configuration"

The idea is to be able to identify the error message easily even for a non-English speaker, or for an English speaker by seeing an error in a different language or with not descriptive content.

Just an idea.

**Comment by bartlomieju** [comment]
We already have that in `DenoError`

**Comment by rsp** [comment]
@bartlomieju do you think it would be a good idea to expose those DenoError codes in messages when presented to the user?

**Comment by bartlomieju** [comment]
> @bartlomieju do you think it would be a good idea to expose those DenoError codes in messages when presented to the user?

I don't have strong opinion about that, current solution has been working fine for me.

You can always check the code yourself like here:
https://github.com/denoland/deno/blob/8d977d0117c2b61e6714cec9e4238f4ba0a56195/cli/js/net_test.ts#L14-L27

**Comment by rsp** [comment]
Actually this example tests for Deno.ErrorKind.Other and checks the message string (which was exactly what hit me when I found out that built-in `Deno.chdir()` throws with a different string for the same error on Windows) but here I was thinking about **user experience of seeing errors** on the terminal and including the codes for ErrorKind there.

By the way, @bartlomieju the current `std/deno` has [_utils.js](https://github.com/denoland/deno/blob/master/std/node/_utils.ts) with `notImplemented()` function to throw errors for unimplemented features that I used. It throws a plain `new Error(message)` instead of `DenoError` - should I change things like this to `DenoError` when I see it? And also if there is not a specific error available should I propose a new one there?

Because using `Deno.ErrorKind.Other` would mean relying on the message strings, defeating the purpose of having error kinds in the first place.

I'd like to update it to be more friendly for people using it and testing for errors (especially since it it a compatibility layer which is incomplete and it will be important to differentiate errors of having the feature unimplemented and having it implemented but failing for some specific reason).

**Comment by bartlomieju** [comment]
The idea is to actually decrease number of those custom errors: https://github.com/denoland/deno/issues/2635

**Comment by rsp** [comment]
Ok, I see. Should std modules even throw those errors or keep using generic `new Error()`?

**Comment by kitsonk** [comment]
Having a good error handling strategy/convention seems important for `std`.  I think there is a difference between a `DenoError` and errors that originate from code in `std`.  Having an error library as part of `std` that provides classes of errors that have more semantic meaning for a given context would be a good idea, but the method should not be fragile (like trying to parse out semantic meaning out of the error message string).

In oak, I adapted [HttpError](https://github.com/oakserver/oak/blob/master/httpError.ts) which allows HTTP errors to be raised in a way they can be responded to and classified.  If we adopted that into `std` and created other classes of errors (we already have `AssertionError`) that provide meaningful semantic extensions to the `Error` class, would be the best way to go about it, IMO.
